pkg/cluster/internal/create/create.go

```diff
@@ -49,66 +57,43 @@ func Cluster(c *context.Context, cfg *config.Config, opts *Options) error {
 		return err
 	}
 
-	// derive info necessary for creation
-	derived, err := Derive(cfg, c.Name())
-	if err != nil {
-		return err
-	}
-
-	// init the create context and logging
-	// TODO(bentheelder): eliminate this
-	cc := &Context{
-		Config:        cfg,
-		DerivedConfig: derived,
-		Context:       c,
-	}
-
-	cc.Status = logutil.NewStatus(os.Stdout)
-	cc.Status.MaybeWrapLogrus(log.StandardLogger())
-
-	defer cc.Status.End(false)
-
-	// TODO(bentheelder): eliminate create context
-	if opts.Retain {
-		cc.Retain = true
-	}
-	if opts.WaitForReady != time.Duration(0) {
-		cc.ExecOptions = append(cc.ExecOptions, WaitForReady(opts.WaitForReady))
-	}
+	status := logutil.NewStatus(os.Stdout)
+	status.MaybeWrapLogrus(log.StandardLogger())
 
 	// attempt to explicitly pull the required node images if they doesn't exist locally
 	// we don't care if this errors, we'll still try to run which also pulls
-	ensureNodeImages(cc.Status, cfg)
+	ensureNodeImages(status, cfg)
 
 	// Create node containers implementing defined config Nodes
-	nodes, err := provisionNodes(cc.Status, cfg, c.Name(), c.ClusterLabel())
+	_, err := provisionNodes(status, cfg, ctx.Name(), ctx.ClusterLabel())
 	if err != nil {
 		// In case of errors nodes are deleted (except if retain is explicitly set)
 		log.Error(err)
-		if !cc.Retain {
-			delete.Cluster(c)
+		if !opts.Retain {
+			delete.Cluster(ctx)
 		}
 		return err
 	}
-	cc.Status.End(true)
-
-	// After creating node containers the Kubernetes provisioning is executed
-	// By default `kind` executes all the actions required to get a fully working
-	// Kubernetes cluster; please note that the list of actions automatically
-	// adapt to the topology defined in config
-	// TODO(fabrizio pandini): make the list of executed actions configurable from CLI
-	err = cc.Exec(nodes, []string{"haproxy", "config", "init", "join"}, cc.ExecOptions...)
-	if err != nil {
-		// In case of errors nodes are deleted (except if retain is explicitly set)
-		log.Error(err)
-		if !cc.Retain {
-			delete.Cluster(c)
+
+	// TODO(bentheelder): make this controllable from the command line?
+	actionsToRun := []actions.Action{
```

Comment:
if each action has a string representation we can have them controllable both from the config and the command line. since it's a fairly advanced feature config only sounds good too.
this would be great!

Reply:
I would like this at some point once the action abstraction is more mature.

In particular it should also be fully re-entrant so you can resume provisioning (IE similar to kubeadm phases)

Reply:
(did not mean to mark this resolved!)